### PR TITLE
Refactoring the project to make it compatible with maven central

### PR DESCRIPTION
**Pull request by jmorra**
The goal of this pull request is to refactor this project to make it compatible with release into [Maven Central](http://search.maven.org/).  In order to do this I've done two things.  First, I've updated the pom file to reflect the requirements outlined [here](http://central.sonatype.org/pages/apache-maven.html) for Maven,  and also added [Travis](https://docs.travis-ci.com/user/languages/java) support.  I have also completely refactored the tests.  I'm not sure how the tests were working before, but judging by the [output](https://ci.appveyor.com/project/lhoang29/explore-java) from AppVeyor they didn't seem to be doing anything.  I would like to advocate for a careful review of what I have written for the new tests.  I'm not really sure what the best way to test this code is, but I took my best guess in some places and I welcome all feedback, but especially surrounding the tests.

In order to actually release this project to Sonatype, someone from MWT has to create an account [here](https://issues.sonatype.org/secure/Signup!default.jspa) and follow the instructions [here](http://central.sonatype.org/pages/ossrh-guide.html) to create a new ticket and request the appropriate Maven coordinates.  If you would like my support on this, please add me (jmorra) as an authorized user and I can push release builds as well.  You may also want to add @deaktator's user name (he can tell you what it is).  Once this is done we can release this project to Maven central so that anyone can easily use it.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1,62 +1,209 @@
...
+
+    <licenses>
+        <license>
+            <name>Multi World Testing License</name>
+            <url>https://github.com/JohnLangford/vowpal_wabbit/blob/master/LICENSE</url>
+        </license>
+    </licenses>
+
+    <developers>
+        <developer>
```
Comment (jmorra) [inline pom.xml:41]: Please feel free to change this to the authors you would like listed.

File: pom.xml [hunk 2]
```diff
@@ -1,62 +1,209 @@
...
+        </snapshotRepository>
+        <repository>
+            <id>ossrh</id>
+            <url>https://oss.sonatype.org/service/local/staging/deploy/maven2/</url>
+        </repository>
+    </distributionManagement>
+
+    <licenses>
+        <license>
+            <name>Multi World Testing License</name>
```
Comment (jmorra) [inline pom.xml:35]: Please suggest the appropriate licensing URL.

**Comment by eisber** [comment]
Looks good to me! Nice work.

File: pom.xml (hunk 1 above)
Comment (lhoang29) [inline pom.xml:41]: I think we can add yourself plus the people listed here: https://github.com/multiworldtesting/decision/blob/master/ClientDecisionService/ClientDecisionService.nuspec#L8

File: pom.xml (hunk 2 above)
Comment (lhoang29) [inline pom.xml:35]: Let's use this one here too: https://github.com/multiworldtesting/decision/blob/master/ClientDecisionService/ClientDecisionService.nuspec#L11

File: src/main/java/com/mwt/tests/BlackBox.java [hunk 3]
```diff
@@ -1,545 +0,0 @@
-package com.mwt.tests;
```
Comment (lhoang29) [inline src/main/java/com/mwt/tests/BlackBox.java:1]: The unit tests below look great, thanks for adding them. However, we should keep the BlackBox tests and have it run in the build server. The point of these tests is to make sure any language-specific interface produces same output as other versions of the exploration libraries. For example, our [C++](https://github.com/multiworldtesting/explore-cpp) and [C#](https://github.com/multiworldtesting/explore-csharp) code are already doing this in appveyor to make sure their exploration is correct. I previously wanted to set this up for Java but haven't got around to it. Basically, the steps needed in the build server are:
1. Clone the [BlackBox test project](https://github.com/multiworldtesting/explore-test) and either the C++ or C# version and build them. For example, the appveyor script for C# is [here](https://github.com/multiworldtesting/explore-csharp/blob/master/appveyor.yml).
2. Trigger the BlackBox test project which will automatically run & compare outputs. This would require that I modify the test project to include the Java BB interface. I'll try to do this but might be a bit later this week as we're in crunch time right now for MWT system. I don't think this should block you from releasing to Maven though because I've manually verified that the results match previously and the core code is unchanged.

One more thing, now that we have Travis set up, we should probably keep BB tests running in both servers. A downside (which I just realized) is that the BB test project is written in C# so we may not be able to run it in Travis. Do you have any suggestions? I'm ok with just running tests in Appveyor for now as well.

**Comment by lhoang29** [comment]
Looks good overall, can you add back the BlackBox test class? I'll merge it in afterwards and set up BB tests automation as mentioned above.

File: src/test/java/com/mwt/explorers/BootstrapExplorerTest.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static junit.framework.Assert.assertEquals;
+
+/**
+ * Created by jmorra on 2/29/16.
+ */
+public class BootstrapExplorerTest {
```
Comment (deaktator) [inline src/test/java/com/mwt/explorers/BootstrapExplorerTest.java:18]: Should these tests have a `RunWith` annotation, as in: 

``` java
@RunWith(BlockJUnit4ClassRunner.class)
```

?

**Comment by lhoang29** [comment]
I created a [ticket on Sonatype](https://issues.sonatype.org/browse/OSSRH-20924). Does it look correct? I tried to add your usernames but it didn't seem to resolve them. @jmorra and @deaktator, what are your Sonatype usernames? I also added both of you as contributors on this repo.

**Comment by deaktator** [comment]
@jmorra, this looks great.  I saw @lhoang29 has a desire to hold onto the black-box tests for cross-language testing.  This is totally fine but I would recommend placing that in `src/test` and exposing a `tests` jar so that the test jar can be used for black-box testing.  For instance, using:

``` xml
<plugins>
  <plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-jar-plugin</artifactId>
    <version>2.4</version>
    <executions>
      <execution>
        <goals>
          <goal>test-jar</goal>
        </goals>
      </execution>
    </executions>
  </plugin>
</plugins>
```

The reason for this is that we don't want to bring in dependencies only relevant to `test` into the _compile-time_ dependencies.  Making the library require fewer dependencies (_actually none_) will make **MWT** easier to use because it will be more lightweight and users won't have to contend with possible version conflicts in downstream projects.

**Comment by deaktator** [comment]
@lhoang29, my sonatype username is **deaktator**.

File: src/test/java/com/mwt/explorers/BootstrapExplorerTest.java (hunk 4 above)
Comment (jmorra) [inline src/test/java/com/mwt/explorers/BootstrapExplorerTest.java:18]: Why is this necessary?  According to [this](http://www.mscharhag.com/java/understanding-junits-runner-architecture) `BlockJUnit4ClassRunner` is the default runner.

Comment (deaktator) [inline src/test/java/com/mwt/explorers/BootstrapExplorerTest.java:18]: It can help with IDE support for junit testing.  For instance, some IDEs have multiple junit versions baked in (e.g., 3 and 4).  Explicitly specifying a version takes away the ambiguity for the IDEs.  I've seen IntelliJ barf in the past one this issue if the IntelliJ configuration is wonky.  This isn't obviously necessary but a nice-to-have.

File: pom.xml (hunk 1 above)
Comment (jmorra) [inline pom.xml:41]: Done.  As for emails, I put everyone's email as mwtdev@microsoft.com except for myself and @JohnLangford.

File: pom.xml (hunk 2 above)
Comment (jmorra) [inline pom.xml:35]: Done

**Comment by jmorra** [comment]
I've taken everyone's suggestions and done the following.
1.  Updated the pom with the appropriate authors
2.  Refactored my tests to make them more reusable.
3.  Added back the BlackBox tests as @lhoang29 suggested.  I combined this with @deaktator's suggestion because I agree with him that it's important to be able to keep the library very lightweight and having no dependencies is awesome.  I've moved the BlackBox tests into another package called `externalTests` and added a `package-info.java` file in there to comment about what @lhoang29 said to alleviate future complications.  In order to run the tests externally simply build the project and issue the following command `java -cp target/explore-java-1.0.0-SNAPSHOT-tests.jar:target/explore-java-1.0.0-SNAPSHOT.jar com.mwt.externalTests.BlackBox`.  Obviously in future versions the 1.0.0-SNAPSHOT should be replaced with the appropriate version.
4.  I've also verified that the Sonatype ticket was properly set up because I was able to issue a `maven clean deploy` and now the SNAPSHOT artifact is up on Sonatype [here](https://oss.sonatype.org/content/repositories/snapshots/com/github/multiworldtesting/explore-java/1.0.0-SNAPSHOT/).

If everyone is happy with this I think it can be merged back into master.  After the merge also make sure to enable the Travis build in the main account and update the README so the Travis badge isn't linked to my Travis account.  After that we should be ready to do the first official deploy the Sonatype which will then automatically promote the artifact to maven central.

The last decision that needs to me made is how to control performing releases.  I would suggest that someone at MWT set up a Jenkins server to handle the release process, although this is not required because releasing should be a one line command, but it's still not a bad idea.  For the first release I'll do it from the command line to make sure everything is working fine and then MWT can take it over.

What does everyone think?

**Comment by lhoang29** [comment]
Looks good to me, merged (assuming others feel the same).
